test/Interop/Cxx/class/constructors-ir.swift

```diff
@@ -0,0 +1,42 @@
+// Target-specific tests for C++ constructor call code generation.
+
+// RUN: %swift -module-name Swift -target x86_64-apple-macosx10.9 -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=ITANIUM_X64
+// RUN: %swift -module-name Swift -target armv7-none-linux-androideabi -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=ITANIUM_ARM
+// RUN: %swift -module-name Swift -target x86_64-unknown-windows-msvc -dump-clang-diagnostics -I %S/Inputs -enable-cxx-interop -emit-ir %s -parse-stdlib -parse-as-library -disable-legacy-type-info | %FileCheck %s -check-prefix=MICROSOFT_X64
+
+import Constructors
+
+typealias Void = ()
+
+public func createHasVirtualBase() -> HasVirtualBase {
+  // - The `this` parameter should carry a `noalias` attribute, as it is
+  //   guaranteed that nothing will alias the object before it has been fully
+  //   constructed. It should also carry an `sret` attribute to indicate that
+  //   this is an out parameter for a structure being returned by the function.
+  //   Note that this doesn't apply on ABIs (Itanium ARM, Microsoft x64)
+  //   where we insert an (inlined) thunk; in this case, we're getting the
+  //   attributes of the constructor that was generated by Clang, which doesn't
+  //   insert these attributes.
+  //
+  // - The `this` parameter should _not_ carry a `nocapture` attribute (unlike
+  //   Swift constructors that return their result indirectly) because the C++
+  //   constructor has explicit access to `this` and may capture it.
+  //
+  // ITANIUM_X64: define swiftcc void @"$ss20createHasVirtualBaseSo0bcD0VyF"(%TSo14HasVirtualBaseV* noalias nocapture sret %0)
+  // ITANIUM_X64-NOT: define
+  // ITANIUM_X64: call void @_ZN14HasVirtualBaseC1E7ArgType(%struct.HasVirtualBase* noalias sret %{{[0-9]+}}, i32 %{{[0-9]+}})
+  //
+  // ITANIUM_ARM: define protected swiftcc void @"$ss20createHasVirtualBaseSo0bcD0VyF"(%TSo14HasVirtualBaseV* noalias nocapture sret %0)
+  // To verify that the thunk is inlined, make sure there's no intervening
+  // `define`, i.e. the call to the C++ constructor happens in
+  // createHasVirtualBase(), not some later function.
+  // ITANIUM_ARM-NOT: define
+  // Note `this` return type.
+  // ITANIUM_ARM: call %struct.HasVirtualBase* @_ZN14HasVirtualBaseC1E7ArgType(%struct.HasVirtualBase* %{{[0-9]+}}, [1 x i32] %{{[0-9]+}})
+  //
+  // MICROSOFT_X64: define dllexport swiftcc void @"$ss20createHasVirtualBaseSo0bcD0VyF"(%TSo14HasVirtualBaseV* noalias nocapture sret %0)
+  // MICROSOFT_X64-NOT: define
+  // Note `this` return type and implicit "most derived" argument.
+  // MICROSOFT_X64: call %struct.HasVirtualBase* @"??0HasVirtualBase@@QEAA@UArgType@@@Z"(%struct.HasVirtualBase* %{{[0-9]+}}, i32 %{{[0-9]+}}, i32 1)
+  return HasVirtualBase(ArgType())
+}
```

Comment:
Could you add some more tests, for cases that don't involve virtual base classes?

For example, you mentioned that constructors on some Itanium ABI variants on ARM64 iOS return 'this', so they would also need a thunk.

Also would be nice to have some tests for cases that don't need thunks.

Feel free to add some of them here if you are not using standard library types -- or in a separate file if you need standard library types.

Reply:
> Could you add some more tests, for cases that don't involve virtual base classes?

Done.

> For example, you mentioned that constructors on some Itanium ABI variants on ARM64 iOS return 'this', so they would also need a thunk.

Some of the existing test cases already demonstrate this -- see the comments on the `ITANIUM_ARM` and `MICROSOFT_X64` tests, which point out the `this` return type.

> Also would be nice to have some tests for cases that don't need thunks.

This was already the case for the existing `ITANIUM_X64` test. The newly added `ITANIUM_X64` test also does not need a thunk.